language-examples/remedial-js/9.0-ES5-vs-ES6/README.md

```diff
@@ -0,0 +1,943 @@
+
+Computed property names allow you to specify properties in object literals based on expressions:
+
+ES6:
+
+```js
+var prefix = 'foo';
+var myObject = {
+  [prefix + 'bar']: 'hello',
+  [prefix + 'baz']: 'world'
+};
+
+console.log(myObject['foobar']);
+// -> hello
+console.log(myObject['foobaz']);
+// -> world
+```
+
+ES5:
+
+```js
+'use strict';
+
+var prefix = 'foo';
+var myObject = {};
+
+myObject[prefix + 'bar'] = 'hello';
+myObject[prefix + 'baz'] = 'world';
+
+console.log(myObject['foobar']);
+// -> hello
+console.log(myObject['foobaz']);
+// -> world
+```
+
+## Destructuring Assignment
+
+The destructuring assignment syntax is a JavaScript expression that makes it possible to extract data from arrays or objects using a syntax that mirrors the construction of array and object literals.
+
+
+ES6:
+
+```js
+var {foo, bar} = {foo: 'lorem', bar: 'ipsum'};
+// foo => lorem and bar => ipsum
+```
+
+ES5:
+
+```js
+'use strict';
+
+var _ref = { foo: 'lorem', bar: 'ipsum' };
+
+// foo => lorem and bar => ipsum
+var foo = _ref.foo;
+var bar = _ref.bar;
+```
+
+ES3:
+
+```js
+with({foo: 'lorem', bar: 'ipsum'}) {
+  // foo => lorem and bar => ipsum
+}
+```
+
+ES6:
+
+```js
+var [a, , b] = [1,2,3];
+```
+
+ES6 (shimming using `Symbol.iterator`):
+
+```js
+'use strict';
+
+var _slicedToArray = function (arr, i) {
+  if (Array.isArray(arr)) {
+    return arr;
+  } else {
+    var _arr = [];
+
+    for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
+      _arr.push(_step.value);
+
+      if (i && _arr.length === i) {
+        break;
+      }
+    }
+
+    return _arr;
+  }
+};
+
+var _ref = [1, 2, 3];
+
+var _ref2 = _slicedToArray(_ref, 3);
+
+var a = _ref2[0];
+var b = _ref2[2];
+```
+
+ES5:
+
+```js
+String.prototype.asNamedList = function () {
+  return this.split(/\s*,\s*/).map(function (name, i) {
+    return name ? ('var ' + name + '=slice(' + i + ', ' + (i + 1) + ')[0]') : '';
+  }).join(';');
+};
+
+with([1,2,3]) {
+  eval('a, , b'.asNamedList());
+}
+```
+
+
+## Default Parameters
+
+Default parameters allow your functions to have optional arguments without needing to check arguments.length or check for undefined.
+
+ES6:
+
+```js
+function greet(msg='hello', name='world') {
+  console.log(msg,name);
+}
+
+greet();
+// -> hello world
+greet('hey');
+// -> hey world
+```
+
+ES5:
+
+```js
+'use strict';
+
+function greet() {
+  // unfair ... if you access arguments[0] like this you can simply
+  // access the msg variable name instead
+  var msg = arguments[0] === undefined ? 'hello' : arguments[0];
+  var name = arguments[1] === undefined ? 'world' : arguments[1];
+  console.log(msg, name);
+}
+
+function greet(msg, name) {
+  (msg === undefined) && (msg = 'hello');
+  (name === undefined) && (name = 'world');
+  console.log(msg,name);
+}
+
+// using basic utility that check against undefined
+function greet(msg, name) {
+  console.log(
+    defaults(msg, 'hello'),
+    defaults(name, 'world')
```

Comment:
its not, the comment above should be interpreted to to mean "you make your own utility based on your use case" type scenario. Its to show that ES6 has this built but if you want this functionality in ES5 you have to make it yourself.